### PR TITLE
22 implement a linear grid current controller

### LINKED ISSUE #22: Implement a linear grid current controller
Implement a linear grid-current controller for RL-grid.

In the library, you will find a file containing an RL-grid model. Create a simple controller that can control the currents of the grid. The controller shall output the desired converter voltage. The controller is to be implemented in soft4pes/control/lin. In order to get everything working, init.py files have to be updated accordingly.

Computational delay shall not be considered.

Test that the controller works by creating another example file. Use the existing example-file as a base and see how e.g. the references are formed.

### PR DESCRIPTION
**Pull request by Milad-Aalto**
The simple PI current controller has been implemented, and anti-windup and voltage feedforward will be added in the future.
closes #22

### CODE REVIEW

File: soft4pes/control/lin_curr_ctr/lin_curr_ctr.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        self.k_i = self.alpha_c * self.R  # Integral gain
+        self.integral_error_d = 0.0
+        self.integral_error_q = 0.0
+        self.u_km1 = np.array([0, 0, 0])
+        self.i_ref_seq_dq = i_ref_seq_dq
+        self.state_space = SimpleNamespace()
+
+    def __call__(self, sys, conv, t):
+        """
+        Compute the switching state at time t.
```
Comment (artturuusila) [inline soft4pes/control/lin_curr_ctr/lin_curr_ctr.py:40]: Replace with "Perform control."

File: soft4pes/control/lin_curr_ctr/lin_curr_ctr.py
[the remark this answers is not in the document]
Comment (Milad-Aalto) [reply soft4pes/control/lin_curr_ctr/lin_curr_ctr.py:1]: Also, I have changed the _init_.py to match with the new directory.

File: soft4pes/control/lin_curr_ctr/lin_curr_ctr.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
+from types import SimpleNamespace
+import numpy as np
+from soft4pes.utils.conversions import alpha_beta_2_dq, dq_2_abc
+
+
+class CurrentControlPI:
```
[the remark this answers is not in the document]
Comment (Milad-Aalto) [reply soft4pes/control/lin_curr_ctr/lin_curr_ctr.py]: I renamed the class's name and applied this change to all files that call this class.

File: soft4pes/control/lin/lin_curr_ctr.py
Comment (artturuusila) [inline soft4pes/control/lin/lin_curr_ctr.py:1]: Add a short description to the beginning of the file. See other files for examples

File: soft4pes/control/lin/lin_curr_ctr.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            Current time [s].
+
+        Returns
+        -------
+        1 x 3 ndarray of floats
+            Modulating signal.
+        """
+
+        # Get the grid voltage
+        vg = sys.get_grid_voltage(t)
```
Comment (artturuusila) [inline soft4pes/control/lin/lin_curr_ctr.py:82]: Not used anywhere

File: soft4pes/control/lin/lin_curr_ctr.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        theta = np.arctan2(vg[1], vg[0])
+
+        # Get the current in dq frame
+        i_dq = alpha_beta_2_dq(sys.x, theta)
+
+        # Compute the converter voltage reference in dq frame using the PI controller
+        u_c_dq = self.pi_controller(i_dq, i_ref_dq)
+
+        # Transform the converter voltage reference back to abc frame
+        u_c_abc = dq_2_abc(u_c_dq, theta)
```
Comment (artturuusila) [inline soft4pes/control/lin/lin_curr_ctr.py:97]: No need to correct this, take this more as a general note: The code should be and it usually is self-descriptive. I know that in programming courses they teach that everything should be commended, but for example here the function name clearly describes what is happening. Therefore, the line comment is not needed. The same applies to lines 97 and 85 for example. 

In the other hand, is it clear to the reader what happens on line 99 `u_k = u_c_abc / (conv.v_dc / 2)` ?

File: soft4pes/control/lin/lin_curr_ctr.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        return np.clip(u_k, -1, 1)  # Ensure modulating signal within -1 and 1
+
+    def pi_controller(self, i_dq, i_ref_dq):
+        """
+        PI controller in dq frame.
+        
+        Parameters
+        ----------
+        i_dq : ndarray
+            Measured current in dq frame [p.u.].
```
Comment (artturuusila) [inline soft4pes/control/lin/lin_curr_ctr.py:110]: Remove measured

File: soft4pes/control/lin/lin_curr_ctr.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        i_ref_dq : ndarray
+            Reference current in dq frame [p.u.].
+
+        Returns
+        -------
+        ndarray
```
Comment (artturuusila) [inline soft4pes/control/lin/lin_curr_ctr.py:116]: What is the array size?

Same comment for the inputs also

File: soft4pes/control/lin/lin_curr_ctr.py (hunk 3 above)
Comment (Milad-Aalto) [reply soft4pes/control/lin/lin_curr_ctr.py:82]: the vg is used to obtain theta to convert alpha beta to dq and other conversions. Do you have another suggestion?

Comment (artturuusila) [reply soft4pes/control/lin/lin_curr_ctr.py:82]: My mistake, this is ok as it is

File: soft4pes/control/lin/lin_curr_ctr.py (hunk 4 above)
Comment (Milad-Aalto) [reply soft4pes/control/lin/lin_curr_ctr.py:97]: Thanks a lot for your comment.

**Comment by edrispouresmaeil** [comment]
This is fine. Thank you!

**Comment by edrispouresmaeil** [comment]
This is fine.

**Review by edrispouresmaeil — APPROVED** [review]

**Pull request merged**